File: 0146-lru-cache/0146-lru-cache.py
class Node:
    def __init__(self, k, v):
        self.key = k
        self.val = v
        self.prev = None
        self.next = None
          
class LRUCache:

    def __init__(self, capacity: int):
        self.cache = {}
        self.head = Node(-1,-1)
        self.tail = Node(-1,-1)
        # self.head.prev = self.head
        self.head.next = self.tail
        self.tail.prev = self.head
        # self.tail.next = self.tail
        self.cap = capacity

    def get(self, key: int) -> int:
        if key in self.cache: 
            self.remove(self.cache[key])
            self.append(self.cache[key])
            return self.cache[key].val
        return -1

    def put(self, key: int, value: int) -> None:
        if key in self.cache: 
            self.remove(self.cache[key])
            self.append(self.cache[key])
            self.cache[key].val = value
        else:
            if len(self.cache)==self.cap: 
                self.cache.pop(self.head.next.key)
                self.remove(self.head.next)
            self.cache[key] = Node(key,value)
            self.append(self.cache[key])
            
    def append(self, nod):
        nod.prev = self.tail.prev
        nod.next = self.tail
        self.tail.prev.next = nod
        self.tail.prev = nod
        
    def remove(self, nod):
        nod.next.prev = nod.prev
        nod.prev.next = nod.next
        
    

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 0146-lru-cache/0146-lru-cache.cpp
class Node{
public:
    int key,value;
    Node *prev,*next;
    Node(int k, int v, Node* p, Node* n)
        :key(k),value(v),prev(p),next(n)
    {
    }
    
    
};
class LRUCache {
public:
    Node *start,*end;
    unordered_map<int,Node*> mp;
    int size;
    LRUCache(int capacity) {
        start = new Node(-1,-1,NULL,NULL);
        end = new Node(-1,-1,start,NULL);
        start->next = end;
        size = capacity;
    }
    
    int get(int key) {
        if(mp.find(key)==mp.end())
            return -1;
        
        // remove from curr position
        mp[key]->prev->next = mp[key]->next;
        mp[key]->next->prev = mp[key]->prev;
        
        // insert at the head
        mp[key]->next = start->next;
        start->next->prev = mp[key];
        mp[key]->prev = start;
        start->next = mp[key];
        
        return mp[key]->value;
    }
    
    void put(int key, int value) {
        
        if(mp.find(key)!=mp.end()){
            mp[key]->value = value;
            get(key);
            return;
        }
        
        if(mp.size()==size){
            Node* del = end->prev;
            mp.erase(del->key);
            end->prev = del->prev;
            del->prev->next = end;
            delete del;
        }
        
        mp[key] = new Node(key,value,start,start->next);
        start->next->prev = mp[key];
        start->next = mp[key];
        
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */